### ISSUE TITLE
[no-redeclare] does not ignore type+namespace with ignoreDeclarationMerge

### ISSUE THREAD

**Issue by paulius-valiunas** [typescript-eslint/typescript-eslint#2818]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-redeclare": [
      "error",
      {
        "ignoreDeclarationMerge": true,
      }
    ]
  }
}
```

```TS
export type A = string | number;
export namespace A {
}
```

**Expected Result**

No errors

**Actual Result**

`error @typescript-eslint/no-redeclare : 'A' is already defined.` (at namespace declaration)
Since the `ignoreDeclarationMerge` flag allows having a class and a namespace with the same name, I would expect it to allow having a type alias and a namespace too.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.2.0` |
| `@typescript-eslint/parser`        | `4.2.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `12.18.4` |

**Comment by bradzacher** [comment]
This isn't a declaration merge. A declaration merge is where the declarations get merged.

```ts
interface A { a: string }
interface A { b: string }
```

These are merged together into a single interface.

The combinations listed in the readme are the only possible declaration merges.

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-redeclare.md#ignoredeclarationmerge

Your example is a redeclaration, and is ignored on purpose, as mentioned in the readme.

**Comment by paulius-valiunas** [comment]
> 
> 
> This isn't a declaration merge. A declaration merge is where the declarations get merged.
> 
> ```ts
> interface A { a: string }
> interface A { b: string }
> ```
> 
> These are merged together into a single interface.
> 
> The combinations listed in the readme are the only possible declaration merges.
> 
> https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-redeclare.md#ignoredeclarationmerge
> 
> Your example is a redeclaration, and is ignored on purpose, as mentioned in the readme.

The problem is, it's NOT ignored. As per documentation:

When set to true, the rule will ignore declaration merges between the following sets:

- interface + interface
- namespace + namespace
- class + interface
- class + namespace
- class + interface + namespace
- function + namespace

Please explain to me, how is class + interface a declaration merge, but type alias + interface is not? How is it different?

**Comment by bradzacher** [comment]
> The problem is, it's NOT ignored.

Sorry - I wasn't clear in my wording.
The combination is purposely ignored from the option, because it's not a declaration merge.

> how is class + namespace a declaration merge, but type alias + namespace is not? How is it different?

```ts
class A {
    declare property: string;
}

namespace A {
    export const variable: number = 1;
}


const x = A.variable
const y = new A();
const z = y.property;
```
[repl](https://www.typescriptlang.org/play?#code/MYGwhgzhAECC0G8CwAoa7oBMCmowCdtoAHfAe2O3wBcBPALmgmvwEsA7AcwG5UBfVKnZgAttgjEwwIvGRoM2AB7EyNaMDLtm0AG4FWYAEYhsjdgFcRhqtAC80AIy8UAlIJQat1aIrtwAdHpsRiaontq0fuzYAO5wABQAlM7h3gBefrT+pBRUdM5AA)

In this example - the namespace declaration and the class declaration were ***merged*** together.
The variables exported from the namespace are available as static members on the class.
The declarations are now indistinguishable as they have been entirely merged together.

Similarly, when you declaration merge an interface and a class - a similar thing happens.

```ts
interface A {
   interfaceProp: number;
}

class A {
    declare property: string;
}

const x = new A();
x.property;
x.interfaceProp;


class B implements A {
    // Class 'B' incorrectly implements class 'A'. Did you mean to extend 'A' and inherit its members as a subclass?
    //    Type 'B' is missing the following properties from type 'A': interfaceProp, property
}

class C extends A {}

const y = new C();
y.property;
y.interfaceProp;
```
[repl](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgILIN4FgBQz+iSyIQAKUA9gA4BcyIArgLYBG0A3LgL664IA2cAM5C0mXPnwATCALhQUVSlWhgAnnSFgooAOaccPHHwogtyAB7IAvPQgB3NAAoAlAYsA6JdVVr3Hwmh4JHJqA14cORFkACFkYCYqfggmCHBRdGw8SQB6HOQAYUFogHIYkviQBAooBQQwfjV4xOTU9OQo0RLUEo9kABFgKWQ1CgZkVLgQZDAKZAgLSBBh7oqp4dAAC2hgMHiwUVTWaFFhZDhkIQYWToB+CVz8yQAVNRVkMorgQ++hPRntsgYBR+PwKPZ-t4VFAwMAIKIYJQmDM3ihVnRAsQQsoADTIKG+bgRTqFeaLNJSDKYIwmMx7Jq2EAOQquAxqLzKXxsgLgIIkUJUAxAA)

Again - the two are now indistinguishable and inseperable.

----

OTOH - when you redeclare a namespace as a type...

```ts
export type A = string | number;
export namespace A {
    export const variable = 1;
}

const x = A;        // x  === typeof A
type t1 = A;        // t1 === string | number (!== typeof a)
type t2 = typeof A; // t2 === typeof a
```
[repl](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcCCcC8cDOMoCWAdgOZwA+cRArgLYBGwUA3ALABQoksVAhrcBxheAY1QYA3hzgy4XaPBEQieOADdehXvQA2qbAEY27AL4cOSlfBBZ0zWQ7gB6J3BtZM2JCggAzdBzeqDAGtmj2jjIuCKGe2HiEpBRUdIxQcAAUAISeCMjAfnC8AJSB+QgATLZBheHOrjBVcXk+-rwcQA)

The two are not merged. The two are just shadowing one-another.
TS will pick the correct one to use depending on how you are using it.
Which means that unlike for all of declaration merging cases - the two are entirely separable. They are not merged.

**Comment by paulius-valiunas** [comment]
OK, that's a fair point... I see lots of such type+namespace combos used as sort of static method holders for union types, but this usage of class+namespace makes more sense.

**Comment by garronej** [comment]
@bradzacher just reporting my opinion here because I feel that this is a big deal. 
Merging type+namespace is a very powerful feature.
Example:
```typescript
type Shape = Shape.Circle | Shape.Square;
namespace Shape {
    export type Circle = {
        type: "circle";
        radius: number;
    };
    export type Square = {
        type: "circle";
        sideLength: number;
    };
}
```
I am so very sad that this rule made it in `create-react-app` 😞   

And I don't understand how class+namespace cut it for you @paulius-valiunas...
